core/src/main/scala/org/typelevel/ci/CaseFoldedString.scala

```diff
@@ -0,0 +1,162 @@
+package org.typelevel.ci
+
+import cats._
+import cats.kernel.LowerBounded
+import org.typelevel.ci.compat._
+import scala.annotation.tailrec
+
+/** A case folded `String`. This is a `String` which has been converted into a
+  * state which is suitable for case insensitive matching under the Unicode
+  * standard.
+  *
+  * This type differs from [[CIString]] in that it does ''not'' retain the
+  * original input `String` value. That is, this is a destructive
+  * transformation. You should use [[CaseFoldedString]] instead of
+  * [[CIString]] when you only want the case insensitive `String` and you
+  * never want to return the `String` back into the input value. In such cases
+  * [[CaseFoldedString]] will be more efficient than [[CIString]] as it only
+  * has to keep around a single `String` in memory.
+  *
+  * Case insensitive `String` values under Unicode are not always intuitive,
+  * especially on the JVM. There are three character cases to consider, lower
+  * case, upper case, and title case, and not all Unicode codePoints have all
+  * 3, some only have 2, some only 1. For some codePoints, the JRE standard
+  * operations don't always work as you'd expect.
+  *
+  * {{{
+  * scala> val codePoint: Int = 8093
+  * val codePoint: Int = 8093
+  *
+  * scala> new String(Character.toChars(codePoint))
+  * val res0: String = ᾝ
+  *
+  * scala> res0.toUpperCase
+  * val res1: String = ἭΙ
+  *
+  * scala> res0.toUpperCase.toLowerCase == res0.toLowerCase
```

Comment:
Yeah, that's part of why my #232 implementation intentionally bypasses any JRE based string conversion and java.util.Locale stuff. 

For example, from my reading on the Turkic rules around `ı` it sounds like they should apply to some Turkic languages, but not necessarily all `Turkic` languages. [The Unicode standard points out these can include ISO-639-2 `tur` and `aze`](https://www.unicode.org/Public/UCD/latest/ucd/CaseFolding.txt), but [wikipedia also includes `kaz` (Kazakh)](https://en.wikipedia.org/wiki/Dotted_and_dotless_I). I thought it might be best to just directly implement the rules, without using `java.util.Locale` to hint anything, and then let the any users interested in this behavior choose to use it under the assumption they would know better when they _specifically_ they should apply.

Reply:
Oh, I was just pointing out that calling these methods without an explicit Locale is setting a bad example for others.  I wish those overloads were deprecated.

Reply:
Ah, I see. I'll make sure anything like that is updated/removed in the final version.